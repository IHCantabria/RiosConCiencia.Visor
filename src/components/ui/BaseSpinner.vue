<script setup>
import { ref, computed, watch } from "vue";
import { SweetModal } from "sweet-modal-vue-3";
import { useSpinnerStore } from "@/store/spinnerStore";
import SpinnerComponent from "@/components/base/ui/SpinnerComponent.vue";

const spinnerStore = useSpinnerStore();

// DATA
const mlBlocking = ref(true);
const mlHideCloseButton = ref(true);
const mlPulseOnBlock = ref(true);
const mlIcon = ref(null);
const mlWidth = ref(null);
const mlEnableMobileFullscreen = ref(false);
const mlOverlayTheme = ref("light");
const mlModalTheme = ref("light");
const loaderRef = ref(null);

// COMPUTED
const mlTitle = computed(() => {
  return spinnerStore.spinnerConfig && spinnerStore.spinnerConfig.title
    ? spinnerStore.spinnerConfig.title
    : "Recopilando";
});
const mlMode = computed(() => {
  return spinnerStore.spinnerConfig && spinnerStore.spinnerConfig.mode
    ? spinnerStore.spinnerConfig.mode
    : 1;
});
const mlText = computed(() => {
  return spinnerStore.spinnerConfig && spinnerStore.spinnerConfig.text
    ? spinnerStore.spinnerConfig.text
    : "Preparando carga de datos. Por favor, espere...";
});
const mlDisplayText = computed(() => {
  return mlMode.value == 1 ? mlText.value : _taskManagerDisplay();
});
const mlSpinnerColor = computed(() => {
  return spinnerStore.spinnerConfig && spinnerStore.spinnerConfig.color
    ? spinnerStore.spinnerConfig.color
    : "#05346d";
});
const data = computed(() => {
  return spinnerStore.mlTasksConfig;
});

// WATCHERS
watch(
  () => spinnerStore.mlActive,
  (newValue) => {
    if (newValue) {
      loaderRef.value.open();
    } else {
      loaderRef.value.close();
    }
  },
);

// METHODS
const _taskManagerDisplay = () => {
  const taskResult = spinnerStore.mlTasksConfig.tasks.find((task) => {
    return task.status === spinnerStore.spinnerEnums.statusTask.pending;
  });
  if (taskResult == null) {
    return mlText.value;
  } else {
    return taskResult.text;
  }
};
</script>

<template>
  <SweetModal
    ref="loaderRef"
    :title="mlTitle"
    :blocking="mlBlocking"
    :hide-close-button="mlHideCloseButton"
    :pulse-on-block="mlPulseOnBlock"
    :icon="mlIcon"
    :width="mlWidth"
    :enable-mobile-fullscreen="mlEnableMobileFullscreen"
    :overlay-theme="mlOverlayTheme"
    :modal-theme="mlModalTheme"
    class="app-style"
  >
    <div class="loading-panel">
      <div class="loading-icon">
        <SpinnerComponent
          v-if="spinnerStore.mlActive"
          :size="50"
          :border-width="5"
          :border-color="mlSpinnerColor"
          :has-absolute-position="true"
        />
      </div>
      <div class="loading-text">{{ mlDisplayText }}</div>
      <div v-if="mlMode == 3" class="loading-table">
        <table
          :data="spinnerStore.mlTasksConfig"
          no-data-text
          max-items="2"
          aria-label="spinner-table"
        >
          <tr v-for="(tr, indextr) in data.tasks" :key="indextr">
            <th class="loading-table__task" :data="data.tasks[indextr].text">
              {{ data.tasks[indextr].text }}
            </th>
            <td v-if="data.loadingBar" class="loading-table__progress">
              <progress
                v-if="!data.tasks[indextr].completed"
                indeterminate
                color="primary"
              ></progress>
              <progress
                v-else-if="data.tasks[indextr].success"
                :percent="100"
                color="success"
              ></progress>
              <progress v-else :percent="100" color="danger"></progress>
            </td>
            <td v-else class="loading-table__status">
              <vs-icon
                v-if="!data.tasks[indextr].completed"
                icon="polymer"
              ></vs-icon>
              <vs-icon
                v-else-if="data.tasks[indextr].success"
                icon="pet"
              ></vs-icon>
              <vs-icon v-else icon="android"></vs-icon>
            </td>
          </tr>
        </table>
      </div>
    </div>
  </SweetModal>
</template>

<style scoped lang="scss">
.loading-panel {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  padding-bottom: 3%;
  align-items: center;
  border-radius: 20px;
}

.loading-icon {
  grid-column: 1;
  grid-row: 1;
  margin-left: auto;
  margin-right: auto;
  padding: 2%;
}

.loading-text {
  grid-column: 2/5;
  grid-row: 1;
  text-align: left;
  margin: 1% 5%;
  font-size: 16px;
}

.loading-table {
  grid-column: 1/5;
  grid-row: 2;
  margin: 5% 5% 0%;
  width: 90%;
  text-align: left;

  &__task {
    width: 80%;
  }

  &__progress {
    width: 20%;
  }

  &__icons {
    width: 20%;
  }
}

.con-table {
  background: rgb(255 255 255 / 80%);
}

/* stylelint-disable-next-line selector-class-pattern */
.table--tbody-table tr {
  background: rgb(255 255 255 / 80%);
}

.sweet-modal-overlay {
  background: rgb(20 20 20 / 70%);
}

/* Non-mobile styles, 750px breakpoint */

@media only screen and (min-width: app-variables.$breakpoint-movil-to-tablet) {
  .loading-panel {
    grid-template-columns: repeat(5, 1fr);
    padding-bottom: 1%;
  }

  .loading-text {
    grid-column: 2/6;
    margin: 1% 3%;
  }

  .loading-table {
    grid-column: 1/6;

    &__task {
      width: 80%;
    }

    &__progress {
      width: 20%;
    }

    &__icons {
      width: 20%;
    }
  }
}
</style>
