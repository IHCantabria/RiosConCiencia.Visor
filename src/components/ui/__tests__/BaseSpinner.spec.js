import { mount, createLocalVue } from "@vue/test-utils";
import BaseSpinner from "../BaseSpinner.vue";
import Vuesax from "vuesax";
import merge from "lodash.merge";

const localVue = createLocalVue();
localVue.use(Vuesax);

function createWrapper(overrides) {
  const defaultMountingOptions = {
    stubs: ["sweet-modal", "circle-loader", "vs-table"],
    localVue
  };
  return mount(BaseSpinner, merge(defaultMountingOptions, overrides));
}

describe("baseSpinner.vue", () => {
  //UNIT TESTS
  it("mlActive watcher", async () => {
    expect.assertions(1);
    const wrapper = createWrapper({});
    const mockWatcher = jest.fn();
    wrapper.vm.$options.watch.mlActive = mockWatcher;
    wrapper.setData({ mlActive: true });
    expect(wrapper.vm.mlActive).toBe(true);
  });
  it("show spinner when show is called", () => {
    expect.assertions(3);
    const wrapper = createWrapper();
    const mockWatcher = jest.fn();
    wrapper.vm.$options.watch.mlActive = mockWatcher;
    const mockShow = jest.fn(wrapper.setData({ calls: 1, mlActive: true }));
    wrapper.vm.show = mockShow;
    wrapper.vm.show();
    expect(mockShow).toHaveBeenCalled();
    expect(wrapper.vm.calls).toBe(1);
    expect(wrapper.find(".app-style").isVisible()).toBe(true);
  });
  it("hide spinner when hide is called", () => {
    expect.assertions(2);
    const wrapper = createWrapper();
    const mockWatcher = jest.fn();
    wrapper.vm.$options.watch.mlActive = mockWatcher;
    const mockHide = jest.fn(wrapper.setData({ calls: 0, mlActive: false }));
    wrapper.vm.hide = mockHide;
    wrapper.vm.hide();
    expect(mockHide).toHaveBeenCalled();
    expect(wrapper.vm.calls).toBe(0);
  });
  it("call _clearTasks when hide is called and calls is 0", () => {
    expect.assertions(2);
    const _clearTasks = jest.fn();
    const wrapper = createWrapper({ methods: { _clearTasks } });
    const mockWatcher = jest.fn();
    wrapper.vm.$options.watch.mlActive = mockWatcher;
    wrapper.setData({ calls: 0, mlActive: false });
    wrapper.vm.calls = 1;
    wrapper.vm.hide();
    expect(wrapper.vm.calls).toBe(0);
    expect(_clearTasks).toHaveBeenCalled();
  });
  it("delete tasks when _clearTasks is called", () => {
    expect.assertions(2);
    const wrapper = createWrapper();
    wrapper.vm.mlTasksConfig.tasks = ["task1", "task2"];
    expect(wrapper.vm.mlTasksConfig.tasks).toHaveLength(2);
    wrapper.vm._clearTasks();
    expect(wrapper.vm.mlTasksConfig.tasks).toHaveLength(0);
  });
  it("add task", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    wrapper.vm.addTask(
      "indexesNowcast",
      "Calculando Indices. Por favor, espere..."
    );
    expect(wrapper.vm.mlTasksConfig.tasks).toHaveLength(1);
  });
  it("update task", () => {
    expect.assertions(2);
    const wrapper = createWrapper();
    wrapper.vm.updateTask(
      "indexesNowcast",
      wrapper.vm.spinnerEnums.statusTask.completed,
      "Cambio"
    );
    const task = wrapper.vm.mlTasksConfig.tasks[0];
    expect(task.status).toBe(wrapper.vm.spinnerEnums.statusTask.completed);
    expect(task.text).toBe("Cambio");
  });
  it("configure mode", () => {
    expect.assertions(2);
    const wrapper = createWrapper();
    expect(wrapper.vm.spinnerConfig.mode).toBe(1);
    wrapper.vm.configureModeSpinner(2);
    expect(wrapper.vm.spinnerConfig.mode).toBe(2);
  });
  it("return text Display _taskManagerDisplay", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    wrapper.vm.mlTasksConfig.tasks = [
      { text: "test", status: 1 },
      { text: "test2", status: 3 }
    ];
    expect(wrapper.vm._taskManagerDisplay()).toBe("test");
  });
  //SNAPSHOT TESTS
  it("renders correctly in basic mode", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("renders correctly in progress mode", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    wrapper.vm.configureModeSpinner(2);
    expect(wrapper.element).toMatchSnapshot();
  });
  it("renders correctly in subTasks mode", () => {
    expect.assertions(1);
    const wrapper = createWrapper();
    wrapper.vm.configureModeSpinner(3);
    wrapper.vm.addTask(
      "clima",
      "Cargando datos de clima atmosf√©rico. Por favor, espere..."
    );
    expect(wrapper.element).toMatchSnapshot();
  });
});
