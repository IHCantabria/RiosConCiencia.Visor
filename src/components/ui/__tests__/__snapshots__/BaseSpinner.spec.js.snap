// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`baseSpinner.vue renders correctly in basic mode 1`] = `
<sweet-modal-stub
  blocking="true"
  class="app-style"
  hide-close-button="true"
  modal-theme="light"
  overlay-theme="light"
  pulse-on-block="true"
  title="Loading"
>
  <div
    class="loading-panel"
  >
    <div
      class="loading-icon"
    >
      <circle-loader-stub
        color="#1c313a"
        size="50"
        sizeunit="px"
      />
    </div>
     
    <div
      class="loading-text"
    >
      test
    </div>
     
    <!---->
  </div>
</sweet-modal-stub>
`;

exports[`baseSpinner.vue renders correctly in progress mode 1`] = `
<sweet-modal-stub
  blocking="true"
  class="app-style"
  hide-close-button="true"
  modal-theme="light"
  overlay-theme="light"
  pulse-on-block="true"
  title="Loading"
>
  <div
    class="loading-panel"
  >
    <div
      class="loading-icon"
    >
      <circle-loader-stub
        color="#1c313a"
        size="50"
        sizeunit="px"
      />
    </div>
     
    <div
      class="loading-text"
    >
      test
    </div>
     
    <!---->
  </div>
</sweet-modal-stub>
`;

exports[`baseSpinner.vue renders correctly in subTasks mode 1`] = `
<sweet-modal-stub
  blocking="true"
  class="app-style"
  hide-close-button="true"
  modal-theme="light"
  overlay-theme="light"
  pulse-on-block="true"
  title="Loading"
>
  <div
    class="loading-panel"
  >
    <div
      class="loading-icon"
    >
      <circle-loader-stub
        color="#1c313a"
        size="50"
        sizeunit="px"
      />
    </div>
     
    <div
      class="loading-text"
    >
      test
    </div>
     
    <div
      class="loading-table"
    >
      <vs-table-stub
        data="[object Object]"
        max-items="2"
        no-data-text=""
      />
    </div>
  </div>
</sweet-modal-stub>
`;
